### DISCUSSION TITLE
How is this project useful

### DISCUSSION THREAD

**Discussion by almahdi404** [AsifArmanRahman/firebase-rest-api#5] [Q&A] [answered]
I am currently exploring firebase sdk in python

Firebase has an [official sdk](https://github.com/firebase/firebase-admin-python) for python

How is this project useful compared to the official sdk?

**Comment by AsifArmanRahman** [comment, marked as answer]
firebase-admin-python : If you use Firebase Admin SDK then you can have admin access to things without worrying about restrictions and rules on your database. 

firebase-rest-api: If you use this library, then you need to authenticate each time, you try access your data (meaning you can set rules and permissions for your database, storage etc. to restrict it from being accessed via another user). Admin access is also available in this library, but it is optional (for details, click [here](https://firebase-rest-api.readthedocs.io/en/latest/guide/firebase-rest-api.html#admin-based-authentication)).

Note: About setting rules and permission [database](https://firebase.google.com/docs/database/security), [storage](https://firebase.google.com/docs/storage/security).

**Comment by almahdi404** [comment]
Thanks for the explanation.

**Comment by Lxstr** [comment]
Does this mean it is basically a full alternative to [Pyrebase4](https://github.com/nhorvath/Pyrebase4)/[Pirebase](https://github.com/josewails/Pirebase) except with added Firestore ability and optional service account? I've been looking for something which I think this might be it! Amazing!

My current understanding from perspective of serverside sessions flask app:

Current enthusiasm from firebase-admin is absent regarding serverside password auth [firebase/firebase-admin-python/issues/641](https://github.com/firebase/firebase-admin-python/issues/641) and hence if you want to use that admin SDK with firestore you still need to use Pyrebase to authenticate the session unless you want to go down the route of client side FirebaseUI [How-to-use-Firebase-Auth-in-Flask/](http://howto.philippkeller.com/2022/06/16/How-to-use-Firebase-Auth-in-Flask/)

One question, when you say authenticate each time above, do you mean pass along token in the same manner as Pyrebase? 
```
rtdb.child("bonuses")
        .child(session["localId"])
        .push(bonus_data, session["idToken"])
```
I couldn't see anything similar in the docs

**Reply by AsifArmanRahman** [reply]
Hello @Lxstr.

<br>

I believe Pyrebase/Pyrebase4/Pirebase already supported optional service account. But for the Firestore part, yes it has this additional ability, as well as use of some social sign in feature.

<br>

> My current understanding from perspective of serverside sessions flask app:
> 
> Current enthusiasm from firebase-admin is absent regarding serverside password auth [firebase/firebase-admin-python/issues/641](https://github.com/firebase/firebase-admin-python/issues/641) and hence if you want to use that admin SDK with firestore you still need to use Pyrebase to authenticate the session unless you want to go down the route of client side FirebaseUI [How-to-use-Firebase-Auth-in-Flask/](http://howto.philippkeller.com/2022/06/16/How-to-use-Firebase-Auth-in-Flask/)

Not clear what you meant over here.

<br>

> One question, when you say authenticate each time above, do you mean pass along token in the same manner as Pyrebase?

Yes, all methods such as ``get, push ... etc`` accepts a parameter ``token`` just the same way Pyrebase does.

**Reply by Lxstr** [reply]
That's great news. Thank you for your work.

To clarify that paragraph, for a flask app with sessions and firestore the options seem to be:

1. firebase-admin + pirebase/pyrebase4 (serverside auth)
2. firebase-admin + firebase-ui (clientside jwt auth)
3. firebase-rest-api 🚀👌
